cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+* Orderability follows comparability but additionally defines a global sort order between values of different types and is aligned with equivalence regarding the handling of `null` and `NaN` values.
+
+=== Examples
+
+An integer compared to a float
+[source, cypher]
+----
+RETURN 1 > 0.5 // should be true
+----
+
+A string compared to a boolean
+[source, cypher]
+----
+RETURN 'string' <= true // should be null
+----
+
+Ordering values of different types
+[source, cypher]
+----
+UNWIND [1, true, '', 3.14, {}, [2]] AS i
+RETURN i
+  ORDER BY i // should not crash
+----
+
+Filtering distinct values of different types
+[source, cypher]
+----
+UNWIND [[null], [null]] AS i
+RETURN DISTINCT i // should return exactly one row
+----
+
+=== Interaction with existing features
+
+The concept of orderability is used only by `ORDER BY` in Cypher today.
+The concept of comparability is used by the comparisons operators `<`, `>`, `<=`, >=`.
+
+One major goal of our proposal is for equality semantics to align well with comparability.
+The concept of equality is used by the equality operator `=`, the inequality operator `<>`, value joins, and the `IN` operator.
+The concept of equivalence is used by the `DISTINCT` clause modifier and in grouping.
+
+With the proposals made in this CIP, specifically changing equality for lists, the mentioned functionality is going to treat lists containing `null` as unequal, thus potentially filtering out more rows when used in a predicate.
+
+=== Alternatives
+
+Columns in SQL are always have a concrete type. This removes the need to define a global sort order between types. Standard SQL has no support for lists, maps, or graph structures and hence does not need to define semantics for them.
+SQL also treats comparisons involving `null` as returning `null`.
+
+PostgresSQL treats some numerical operations (like division by zero) that would compute a `NaN` as a numerical error that fails the query. PostgresQL considers `NaN` to be larger than positive infinity, both in comparison and in sort order. This proposal achieves something very similar by evaluating comparisons involving a `NaN` to `null` and by treating both `NaN` and `null` as the largest values in the global sort order.
+
+This proposal could be extended with an operator for making equivalence accessible beyond use in grouping and `DISTINCT`. This seems desirable due to plain `=` not being reflexive for all values.
+
+Also noteworthy: If nulls would track their source, equality could become reflexive again as it would become possible to know if two `null` values represent the same "unknown" value.
+
+== Benefits to this proposal
+
+A consistent set of rules for equality, equivalence, comparability and orderability.
```

Comment:
"A consistent set of rules for equality, equivalence, comparability and orderability." 
-->
"A consistent set of rules is defined for equality, equivalence, comparability and orderability."
(Currently a clause rather than a sentence...)